packages/safe-apps-developer-ui/src/stores/provider.ts

```diff
@@ -1,29 +1,53 @@
 import create from 'zustand';
-import { Web3Provider } from '@ethersproject/providers';
+import { Web3Provider, JsonRpcSigner } from '@ethersproject/providers';
 import { ETHEREUM_NETWORK_TO_ID, WALLET_PROVIDER } from 'src/api/provider';
 
-type ProviderState = {
-  loaded: boolean;
-  account: string;
-  name: string;
-  networkId: ETHEREUM_NETWORK_TO_ID;
+type ProviderInfo = { loaded: boolean; account: string; name: string; networkId: ETHEREUM_NETWORK_TO_ID };
+
+type ProviderState = ProviderInfo & {
   provider: Web3Provider | null;
+  signer: JsonRpcSigner | null;
   fetchAndSetProvider: (provider: Web3Provider) => Promise<void>;
   disconnect: () => void;
+  updateProvider: () => void;
 };
 
-const useProviderStore = create<ProviderState>((set) => ({
+const useProviderStore = create<ProviderState>((set, get) => ({
   loaded: false,
   account: '',
   name: WALLET_PROVIDER.UNKNOWN,
   networkId: ETHEREUM_NETWORK_TO_ID.UNKNOWN,
   provider: null,
+  signer: null,
 
   fetchAndSetProvider: async (provider: Web3Provider) => {
     const account = (await provider.listAccounts())[0];
     const { chainId: networkId } = await provider.getNetwork();
 
-    return set({ account, loaded: true, networkId, provider });
+    return set({ account, loaded: true, networkId, provider, signer: provider.getSigner() });
+  },
+
+  updateProvider: async () => {
+    const { provider } = get();
+
+    if (!provider) {
+      return;
+    }
+
+    const account = (await provider.listAccounts())[0];
+    const { chainId: networkId } = await provider.getNetwork();
+
+    if (!account) {
+      return set({
+        loaded: false,
+        account: '',
+        networkId: ETHEREUM_NETWORK_TO_ID.UNKNOWN,
+        provider: null,
+        signer: null,
+      });
+    }
+
+    return set({ account, loaded: true, networkId });
```

Comment:
Should the `provider` and `signer` be also added here?

Reply:
It will partially update the state, we don't need to update them

Reply:
What if the provider is updated twice?
- The first update has no account and the provider is initialized with all the properties to their default values (null, false, etc.)
- The second update has an account and the properties `account`, `loaded` and `networkId` are updated but `provider` and `signer` are still null.

As far as I understand.

Reply:
It's only invoked when there's an existing provider and account/network changes: https://github.com/gnosis/safe-apps-sdk/pull/135/files#diff-96c3a90c6ed644ae146525eedc19cc94ac0471e7674b81c28cb24d34b4b46cf6R29-R30